File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        diff, sum_ = float("inf"), None
        for i in range(0, len(nums)):
            l, r = i + 1, len(nums)-1
            while l < r:
                cur_sum = nums[i] + nums[l] + nums[r]
                cur_diff = abs(target - cur_sum)
                if cur_diff == 0:
                    return cur_sum
                elif cur_diff < diff:
                    diff = cur_diff
                    sum_ = cur_sum
                elif cur_sum > target:
                    r -= 1
                else:
                    l += 1
        return sum_

File: 0016-3sum-closest/0016-3sum-closest.ts
function threeSumClosest(nums: number[], target: number): number {
    nums.sort((a, b) => a - b)
    return kSumClosest(3, 0, target)
    
    function kSumClosest(k: number, i: number, target: number): number {
        if (k === 2) return twoSumClosest(i, target)
        let diff = Infinity
        let sum: number
        for (i; i < nums.length; i++) {
            if (i > 0 && nums[i] === nums[i-1]) continue
            const kMinusOneSumClosest = kSumClosest(k-1, i+1, target-nums[i])
            const curSum = nums[i] + kMinusOneSumClosest
            const curDiff = Math.abs(target - curSum)
            if (curDiff === 0) {
                return curSum
            } else if (curDiff < diff) {
                diff = curDiff
                sum = curSum
            }
        }
        return sum
    }
    
    function twoSumClosest(i: number, target: number): number {
        let l = i
        let r = nums.length-1
        let diff = Infinity
        let sum: number
        while (l < r) {
            const curSum = nums[l] + nums[r]
            const curDiff = Math.abs(target - curSum)
            if (curDiff === 0) {
                return curSum
            } else if (curDiff < diff) {
                diff = curDiff
                sum = curSum
            } else if (curSum > target) {
                r--
            } else {
                l++
            }
        }
        return sum
    }
};



